### PR TITLE
Fixed REST API: verify values of parameters skip and limit #2913

### LINKED ISSUE #2913: REST API: verify values of parameters skip and limit
Currently, when parameters *skip* and *limit* are provided, in program code their values are converted to integer. Unless the value is not integer, the given value is ignored.

The provided value of parameter lifecycle is not validated.

The values need to be validated and an error message must be returned in case values are not feasible.

### PR DESCRIPTION
**Pull request by deligence-dharmendra**
Closes #2913

## Changes

- changes in server api js file to restrict parameters data type

## Developer checklist
This checklist is to be completed by the PR developer:

- [ ] Alternative solutions were compared/discussed before writing code
    - [ ] trade-offs with this solution are considered acceptable
- [ ] Code in this PR adheres to the [project styleguide](CONTRIBUTING.md)
- [ ] This pull request does not decrease project test coverage
- [ ] If the code changes existing database collection(s), migration has been written
- [ ] If UI texts are added or changed, all texts are internationalized

## Reviewer checklist
Reviewed by: @matleppa  @marla-singer 

This list is to be completed by the pull request reviewer:
- [ ] Code works as described/expected
- [ ] Code seems to be error free
    - [ ] no browser console errors visible
    - [ ] no server console errors visible
    - [ ] passes CI build
- [ ] Code is written in a way that promotes maintainability
    - [ ] easy to understand
    - [ ] well organized
    - [ ] follows project coding standards and conventions
    - [ ] well documented

### CODE REVIEW

File: apinf_packages/apis/server/api.js [hunk 1]
```diff
@@ -114,14 +114,29 @@ CatalogV1.addCollection(Apis, {
...
+          if (!queryParams.lifecycle.match(/^[a-z]+$/)) {
+            return errorMessagePayload(400,
+              'Bad query parameters value. Lifecycle parameters only accept string.');
+          }
           query.lifecycleStatus = queryParams.lifecycle;
         }
 
         if (queryParams.limit) {
+          // Make sure limit parameters only accept integer
+          if (!queryParams.limit.match(/^\d+$/)) {
```
Comment (marla-singer) [inline apinf_packages/apis/server/api.js]: Don't use the regular expression for checking the simple thing. 
Use function `Number.isInteger` 
https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger

File: apinf_packages/apis/server/api.js [hunk 2]
```diff
@@ -114,14 +114,29 @@ CatalogV1.addCollection(Apis, {
...
+            return errorMessagePayload(400,
+              'Bad query parameters value. Limit parameters only accept integer.');
+          }
           options.limit = parseInt(queryParams.limit, 10);
         }
 
         if (queryParams.skip) {
+          // Make sure skip parameters only accept integer
+          if (!queryParams.skip.match(/^\d+$/)) {
```
Comment (marla-singer) [inline apinf_packages/apis/server/api.js]: Use function `Number.isInteger` 
https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger

File: apinf_packages/apis/server/api.js [hunk 3]
```diff
@@ -114,14 +114,29 @@ CatalogV1.addCollection(Apis, {
         }
 
         if (queryParams.lifecycle) {
+          // Make sure lifecycle parameters only accept string
+          if (!queryParams.lifecycle.match(/^[a-z]+$/)) {
```
Comment (marla-singer) [inline apinf_packages/apis/server/api.js]: The same stuff is here. Don't use the regular expression for checking the simple thing.
For your example, you can check if a parameter is String type using another way
https://stackoverflow.com/questions/4059147/check-if-a-variable-is-a-string-in-javascript

In this case, `lifecycle` parameter is expected to be one of the list values ['design', 'development', 'testing', 'production', 'deprected']

**Comment by deligence-dharmendra** [comment]
Hi @marla-singer **typeof** not properly validate string value in our case. If someone used lifecycle=1234. **typeof** treated 1234 as a string.In my knowledge we have two options.Either we go for regular expression or go for str.charCodeAt like
```
 for (i = 0, len = str.length; i < len; i++) {
    code = str.charCodeAt(i);
    if (!(code > 96 && code < 123)) { // lower alpha (a-z)
      return false;
    }
  }
```
Please advise.

**Comment by marla-singer** [comment]
@deligence-dharmendra  As I said, in our case, lifecycle parameter must be **be one of the list values** ['design', 'development', 'testing', 'production', 'deprected']. Check if the lifecycle parameter is included in the list

**Comment by matleppa** [comment]
@deligence-dharmendra 

Here is the field in question in apis collection.

```
 lifecycleStatus: {
    type: String,
    optional: true,
    allowedValues: [
      'development',
      'deprecated',
      'design',
      'production',
      'testing',
    ],
  },

```
If the query parameter for lifecycle status is given in /GET method, in has to be one of those allowed values, as @marla-singer said.
However, I think that we can show some flexibility here and make lifecycle parameter in query case insensitive.
```
IF query parameter for lifecyclestatus is given
  make it lower case
  validate it against the list of allowed values
  IF not match
    return error
  ELSE 
    use lower-cased parameter in query


```
And I think that the validation function has to be used here so the functionality is up-to-date in case the list of allowed values change in future.

**Pull request merged**